doc/design/resource-dependency.md

```diff
@@ -0,0 +1,343 @@
+      Resource ResourceReference `json:"resource"`
+
+      // FieldPath specifies the field path of a given k8s-style resource.
+      FieldPath string `json:"fieldPath"`
+    }
+
+    type ResourceReference struct {
+      APIVersion string `json:"apiVersion"`
+
+      Kind string `json:"kind"`
+
+      // Name indicates the name of the object in the same namespace.
+      Name string `json:"name,omitempty"`
+    }
+
+    type DataOutputStatus struct {
+      // Ready indicates whether the field data of the source resource is ready.
+      Ready bool `json:"ready"`
+    }
+    ```
+
+    For example, a mysql connection secret DataOutput coming from RDSInstance my-rds would look like:
+
+    ```yaml
+    apiVersion: experimental.core.oam.dev/v1alpha2
+    kind: DataOutput
+    metadata:
+      name: mysql-conn-secret
+    spec:
+      resource:
+        apiVersion: alibaba.oam.crossplane.io/v1alpha1
+        kind: RDSInstance
+        name: my-rds
+      fieldPath: "status.connSecret"
+    ```
+
+    In this case, the OAM runtime will automatically build a dependency between the `my-web` and `my-rds`.
+
+    It will wait until `my-rds` object's field `status.connSecret` has value, and marks the DataOutput status ready:
+
+    ```yaml
+    apiVersion: alibaba.oam.crossplane.io/v1alpha1
+    kind: RDSInstance
+    metadata:
+      name: my-rds
+    spec:
+      ...
+    status:
+      connSecret: mysql-conn # OAM runtime will wait for this field to have value
+    ---
+    apiVersion: experimental.core.oam.dev/v1alpha2
+    kind: DataOutput
+    metadata:
+      name: mysql-conn-secret
+    ...
+    status:
+      ready: true # once the above connSecret has value this will be marked true
+    ```
+
+    At this point, the dependency has been satisfied.
+    The OAM runtime will create `my-web` object with corresponding values passed to its fields.
+    
+    ```yaml
+    kind: MyWorkload
+    metadata:
+      name: my-web
+      ..
+    spec:
+      connSecret: mysql-conn # patched by OAM runtime
+    ```
+
+3. Below we show how to combine the above concepts (DataInput, DataOutput) into OAM.
+
+    In OAM Component we put a syntax-sugar on existing `Parameters` fields:
+    as long as the name of the Parameter matches the name of the DataOutput, it becomes a DataInput.
+    For example, the following parameter mysql-conn-secret in Component matches the name of DataOutput trait.
+
+    ```yaml
+    kind: Component
+    metadata:
+      name: my-web
+    spec:
+      workload:
+        kind: MyWorkload
+        metadata:
+          name: my-web
+      parameters:
+      - name: mysql-conn-secret # this matches the name of DataOutput
+        fieldPaths: ["spec.connSecret"]
+    ---
+    kind: Component
+    metadata:
+      name: my-rds
+    spec:
+      workload:
+        kind: RDSInstance
+        metadata:
+          name: my-rds
+    ---
+    # In ApplicationConfiguration
+    components:
```

Comment:
In this case I am not sure how much using the `DataOutput` is helping the end user. They still have to be aware of the connection secret field of the `RDSInstance` which is a level lower than the `Component` they are using. It seems like a better experience would be for the `Component` author to surface a `secretName` parameter for the `my-rds` and `my-web` `Components`, then the end user could just provide the same value for each. The user would only be required to know about the components rather than knowing about the underlying resource types, which seems like the abstraction that we want at the `ApplicationConfiguration` level.

Reply:
Consider the experience here vs this from the quick start with service tracker:

```yaml
spec:
  components:
    - componentName: data-api-database
      parameterValues:
        - name: database-secret
          value: tracker-database-secret
    - componentName: data-api
      parameterValues:
        - name: database-secret
          value: tracker-database-secret
```

Reply:
I totally understand. But this is something we could not enforce users to adopt. For example, in Alibaba cloud, the ID of RDS would not be available at the beginning until RDS instance is created. This has to be passed from the status.